Hook up profiled execution in the interpreter.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21616 [WIP] Hook up profiled execution in the interpreter.**

suo committed:
Consider contained types in alias analysis (#21431)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21431
ghimport-source-id: d86ce974a065ec572e71cfa14a8f6bdf48216da7

Reviewed By: jamesr66a

Differential Revision: D15718560

Pulled By: suo

fbshipit-source-id: a36ce907ab26be22f12bab6175797fe8b34721f1

suo committed:
cleanups to alias analysis interfaces (#21397)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21397
ghimport-source-id: 8733e1af2fe66a3f4494a2c24c82a039375a982e

Reviewed By: jamesr66a

Differential Revision: D15642662

Pulled By: suo

fbshipit-source-id: ae66b7b4f19f255d6fe0e7e804bd0df6d86cb8d1

suo committed:
cleanups to memory_dag (#21430)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21430
ghimport-source-id: 2dc5a0df8512e796c12d65d3ecc5981638122ce6

Reviewed By: jamesr66a

Differential Revision: D15718561

Pulled By: suo

fbshipit-source-id: 1ef31c08c8a757b632451eb07a47a8227e76c67f

suo committed:
avoid calling front() on empty working set (#21396)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21396
ghimport-source-id: 7e57282099d2fd57c58c990b51ae933e427aecb2

Reviewed By: jamesr66a

Differential Revision: D15642663

Pulled By: suo

fbshipit-source-id: f9b467ba53f03438879bf3929da522aabaff2343

suo committed:
fix optional type promotion for classes (#21593)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21593
ghimport-source-id: f68730618bccf2326218e08d0a2a70171fdd8921

Differential Revision: D15741471

Pulled By: suo

fbshipit-source-id: 7ac1a0f6d9d2ff4bc819caff43a7a5b6d37cbc98

colesbury committed:
Revert D15717575: [pytorch][PR] Fix bug in multinomial_alias_draw

Differential Revision:
D15717575

Original commit changeset: b1154e226d42

fbshipit-source-id: 305ca010bfda88c9295c52e0626d867452c72f84

smessmer committed:
Allow tensors with requires_grad=True in c10 ops (#21599)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21599

We prevented this because c10 ops can't have a backwards yet and calling them with requires_grad=True would do the wrong thing
if the c10 op is not purely implemented by calling other autograd-able ops.

However, it is a valid use case to have c10 ops that just call other autograd-aware ops, and these ops should be callable with requires_grad=True.

This should fix https://github.com/pytorch/pytorch/issues/21584.

Differential Revision: D15744692

fbshipit-source-id: ba665365c850ef63fc9c51498fd69afe49e5d7ec

gchanan committed:
Use schema string specification in derivatives.yaml. (#20916)

Summary:
For consistency, derivatives.yaml now uses the same schema specification as native_functions.yaml.

Note that there are some small downsides, e.g. changing the default values or return parameter names in native_functions.yaml also now requires updating derivatives.yaml as well.  But this has a few nice properties:
1) Able to copy-paste definitions from native_functions to derivatives.
2) Makes it impossible to write derivatives for operators without schemas (e.g. old TH operators).
3) Moves us closer to the ideal situation of co-locating forward and backwards declarations.

Note that this doesn't change any generated code; in particular, this has the same behavior of mapping in-place and out-of-place definitions together.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20916

Differential Revision: D15497800

Pulled By: gchanan

fbshipit-source-id: baee5caf56b675ce78dda4aaf6ce6a34575a6432

eellison committed:
Uninitialized Ivalue (#21387)

Summary:
Create an uninitialized ivalue. This will be needed for Breaks & Continues to match up if block outputs of values that are guaranteed not to be used but need to escape the block scope. It is not exposed to users.

Was previously part of final returns but I was asked to make a separate PR for it.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21387

Differential Revision: D15745124

Pulled By: eellison

fbshipit-source-id: ae6a6f766b4a70a71b9033987a630cfbf044e296

pandeykartikey committed:
Implements divmod function (#20979)

Summary:
This PR refer to issue #18627
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20979

Differential Revision: D15743929

Pulled By: wanchaol

fbshipit-source-id: 967fc3fd519501e427176e10b112c8be1390540b

zdevito committed:
[WIP] Hook up profiled execution in the interpreter.

skrah committed:
Add a 'dim' argument to nuclear norm (#21022)

Summary:
Addresses #18275.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21022

Differential Revision: D15743515

Pulled By: ezyang

fbshipit-source-id: e4aaea0bd7f863a2abad45c4322d6a9fb02a88e3

ezyang committed:
Revert "Redefine scheduler to set learning rate using recursive formula" #14010 (#21463)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21463
ghimport-source-id: 1b0ea4a282b41388d5c6f6a5d18d37c14ae874ad

Differential Revision: D15747426

Pulled By: ezyang

fbshipit-source-id: 0708394f907b98a9f45bcfa26e5cc450fda8cf76

PenghuiCheng committed:
support negative axis in concat and split operators

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/17955

Differential Revision: D14476031

Pulled By: ezyang

fbshipit-source-id: e0e57e8595ed2005ded9e923572a40fe62aca5a7

mrshenli committed:
Run test_c10d.py in multi-gpu environment (#21598)

Summary:
yf225 helped me discovered that our CI does not run multi-gpu tests in `test_c10d.py`. There are quite a few multi-gpu c10d tests. This PR tries to enable those tests.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21598

Differential Revision: D15744256

Pulled By: mrshenli

fbshipit-source-id: 0a1524a862946128321f66fc8b7f331eff10e52a

smessmer committed:
Publish torch::Dict and torch::OperatorKernel (#20723)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20723

These classes already existed but only as c10::Dict and c10::OperatorKernel.
Since they're now part of torch::RegisterOperators(), they should also live in the torch namespace.

Differential Revision: D15421575

fbshipit-source-id: d64ebd8664fadc264bbbae7eca1faa182529a32b

mingzhe09088 committed:
add PT gather op to the benchmark suite (#21614)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21614

as title

Reviewed By: kimishpatel

Differential Revision: D15525115

fbshipit-source-id: 6a17e1d791bdb432cc3d51e45c5e82b96268127d

suo committed:
Split out Function into its own file (#21539)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21539
ghimport-source-id: f1e4396a0bec6e30d3179f926ec4da68807942f7

Differential Revision: D15741979

Pulled By: suo

fbshipit-source-id: 4cd0ed36bcbf8db0b36a101dda6f58975f806889

smessmer committed:
Deprecate torch::jit::RegisterOperators (#21552)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21552

Original commit changeset: a142c22be3fd

https://github.com/pytorch/pytorch/pull/21368 got reverted because of a MSVC issue. This commit re-introduces that change and fixes the MSVC issue.

Differential Revision: D15727526

fbshipit-source-id: 8eb0eb9a7108dc049911b79342c364ac1b8623c8

bwasti committed:
AT_CHECK -> TORCH_CHECK (#21547)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21547
ghimport-source-id: d99d3fdcd9abde4e1126716d32ed05aaf8508c50

Differential Revision: D15747676

Pulled By: bwasti

fbshipit-source-id: ae9824436e8316e2d0002d2973df4833a18c5f23

Commit:
Method-only entries in native functions should have self as first argument (#21549)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21549
ghimport-source-id: a98fd7a18b4c523d9facb328a3b80a35416834ce

Differential Revision: D15724794

Pulled By: li-roy

fbshipit-source-id: a0f218cf6fd32d9694921685fc805d868156fce3

mingzhe09088 committed:
add separate path for op with JIT (#21210)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21210

This diff introduces a new path to run op with JIT. There are two steps involved here:
1. Users need to script the op. This should happen in the `init` method.
2. The generated graph from step1 is passed to `jit_forward` which will be executed by the benchmark backend

Reviewed By: zheng-xq

Differential Revision: D15460831

fbshipit-source-id: 48441d9cd4be5d0acebab901f45544616e6ed2ee

mingzhe09088 committed:
add PT pointwise unary ops to the benchmark suite (#21207)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21207

This diff adds 80 PT pointwise unary ops to the benchmark suite. Most of the ops are added using the generate_pt_tests_from_list interface. The rest are handled separately.

Reviewed By: zheng-xq

Differential Revision: D15471597

fbshipit-source-id: 8ea36e292a38b1dc50f064a48c8cd07dbf78ae56

gujinghui committed:
upgrade mkldnn-bridge (#20569)

Summary:
1. reduce the overhead of mkldnn-bridge itself
2. remove redundant code and useless APIs
3. provide new operators, including int8 inner_product,  ND permute/transpose, elem_add/mul, and etc.
4. improve inner_product to support io format weights without implicit reorder
5. add SoftMax support
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20569

Reviewed By: houseroad

Differential Revision: D15558663

Pulled By: bddppq

fbshipit-source-id: 79a63aa139037924e9ffb1069f7e7f1d334efe3a

gujinghui committed:
implement transpose operator for MKLDNN (#19955)

Summary:
implement transpose operator for MKLDNN
1. upgrade mkldnn-bridge to support ND transpose
2. implement transpose operator in caffe2.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19955

Differential Revision: D15701832

Pulled By: bddppq

fbshipit-source-id: e4337cd0ba6f8180a35c8c70cbb6830a0a84182f

zou3519 committed:
Fix namedtensor build (#21609)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21609
ghimport-source-id: 648a0bcd28db2cdda1bf2fa6a904ca8f851088c2

Differential Revision: D15747687

Pulled By: zou3519

fbshipit-source-id: 2a972a15fa7399391617fc6e6b19879b86568c3a

xuhdev committed:
Move management of calls of "cmake --build" to setup_helper/cmake.py and refactor as a CMake class

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21493

Differential Revision: D15759279

Pulled By: ezyang

fbshipit-source-id: 157e1de36f1c5a51caf2a25b363a94369c442012

brianjo committed:
Breaks up NN module in docs so it loads faster.

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21291

Differential Revision: D15760935

Pulled By: ezyang

fbshipit-source-id: 114da4c52b78949e631e9adcae4eb620546124fb

Commit:
better error msg when seeing a unsupported builtin function (#21068)

Summary:
fixes https://github.com/pytorch/lockdown/issues/39.
Hopefully it doesn't break other tests....
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21068

Differential Revision: D15761895

Pulled By: ailzhang

fbshipit-source-id: 60cbb16cfc930b377d753b81e10b7edaea9a1281

mrshenli committed:
Fix Process Group for tensors shared across processes (#21449)

Summary:
Ops on a Process Group (pg) instance will hit an error when input/output tensors are created on a different process, because, pg calls `recordStream` on `CUDACachingAllocator` which only knows tensors created within the same process.

The proposed solution is to add a `suppressError` arg (suggestions for better names?) to `recordStream`. See comments in code for arguments.

CC pichuang1984
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21449

Differential Revision: D15689736

Pulled By: mrshenli

fbshipit-source-id: e7fc81b167868f8666536067eaa7ae2c8584d88e

Commit:
Support for type annotations instead of torch.jit.annotate() (#21390)

Summary:
This adds support for PEP 526 style annotations on assignments in place of
`torch.jit.annotate()`, so

```python
a = torch.jit.annotate(List[int], [])
```

turns into

```python
a : List[int] = []
```
](https://our.intern.facebook.com/intern/diff/15706021/)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21390

Pulled By: driazati

Differential Revision: D15706021

fbshipit-source-id: 8bf1459f229d5fd0e16e59953b9656e85a2207fb

szagoruyko committed:
Rename hubconf.conf to hubconf.py in docs (#21631)

Summary:
It's a typo I guess. cc fmassa
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21631

Differential Revision: D15764909

Pulled By: soumith

fbshipit-source-id: 5ffc7bde181c13e151332e7de3c0da36505b495e

VitalyFedyunin committed:
Fix catch block to fix 'error: catching polymorphic type' (#21637)

Summary:
Fix catch block to fix 'error: catching polymorphic type `class c10::Error` by value [-Werror=catch-value=]'
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21637

Differential Revision: D15761860

Pulled By: VitalyFedyunin

fbshipit-source-id: befc18a9c217440381cdb50a1319b0b5db5710e9

Commit:
Revert D15706021: [jit] Support for type annotations instead of torch.jit.annotate()

Differential Revision:
D15706021

Original commit changeset: 8bf1459f229d

fbshipit-source-id: 7ae34578560e2dccd0f04af2220445b3999771fe

bwasti committed:
Add len to OrderedDict types (#21651)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21651
ghimport-source-id: 0bba5c1930865e2d18b18782ba8c8990b0761d4d

Differential Revision: D15767795

Pulled By: bwasti

fbshipit-source-id: 70e27176897b0f977c9034ffb3ad21091c91e12e

Krovatkin committed:
fixes to make a profiling pipeline work

clean up

bettter tests

address michael's feedback

sink guard creation into the true arm of an if

fix build failures

more build errors after renaming a function

missed one

emerali committed:
Fix momentum bug in CyclicLR (#20401)

Summary:
Resolves issue https://github.com/pytorch/pytorch/issues/19003

The author of this issue also asked that `cycle_momentum` default to `False` if the optimizer does not have a momentum parameter, but I'm not sure what the best way to do this would be. Silently changing the value based on the optimizer may confuse the user in some cases (say the user explicitly set `cycle_momentum=True` but doesn't know that the Adam optimizer doesn't use momentum).

Maybe printing a warning when switching this argument's value would suffice?
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20401

Differential Revision: D15765463

Pulled By: ezyang

fbshipit-source-id: 88ddabd9e960c46f3471f37ea46013e6b4137eaf

vishwakftw committed:
Fix bug in multinomial_alias_draw (#21324)

Summary:
An incorrect increment / decrement caused the samples to not be generated from a multinomial distribution

Changelog:
- Remove the incorrect increment / decrement operation

Fixes https://github.com/pytorch/pytorch/issues/21257, fixes https://github.com/pytorch/pytorch/issues/21508

cc: LeviViana neerajprad
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21324

Differential Revision: D15761029

Pulled By: colesbury

fbshipit-source-id: 2aeb51e2d3cfdb8356806a7d5b12d4b9910e37fb

bwasti committed:
Add python string standard lib (#21656)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21656
ghimport-source-id: cc7d7f68e33e95a97f6274c50823138aa4bacabb

Differential Revision: D15769256

Pulled By: bwasti

fbshipit-source-id: 1af487446361d90d03dce004c3e2169a3e62667d

Commit:
make range functions respect current stream (#21619)

Summary:
Stream is not respected on range/linspace/logspace functions, which contributes to https://github.com/pytorch/pytorch/issues/21589 (this is not a complete solution for that issue).
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21619

Differential Revision: D15769666

Pulled By: ezyang

fbshipit-source-id: 7c036f7aecb3119430c4d432775cad98a5028fa8

dskhudia committed:
randomized testing for qconv (#21436)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21436

Test many different options

Reviewed By: zafartahirov

Differential Revision: D15683754

fbshipit-source-id: 60d0fc697b53c7e4adadbe80995d45f28729bca4

Commit:
test_jit: Remove tests checking non-guaranteed properties from 'test_insert_observers'. (#21657)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21657
ghimport-source-id: e9c7e45c00db55bf3b7895d06d77f0d99bfc1afe

Differential Revision: D15769295

Pulled By: ZolotukhinM

fbshipit-source-id: cfb40bc5d7116b1d99f5e0f5c4f5577f5aa33804

Commit:
fix incorrect use of TeX in docs

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21649

Differential Revision: D15766392

Pulled By: umanwizard

fbshipit-source-id: a362ec06e971ee12c47a45bc9c15cc773ec878e3

Commit:
Revert D15769256: [jit] Add python string standard lib

Differential Revision:
D15769256

Original commit changeset: 1af487446361

fbshipit-source-id: 96bea4a49664dad68762bef75ae28e64c673f8b1

houseroad committed:
Automatic update of fbcode/onnx to dd599b05f424eb161a31f3e059566a33310dbe5e (#21641)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21641

Previous import was 5160f3ac3380302224998f1c95e111cd961c4bc5

Included changes:
- **[dd599b05](https://github.com/onnx/onnx/commit/dd599b05)**: Fix type s/depracted/deprecated/ (#2092) <Takeshi Watanabe>
- **[abb1702a](https://github.com/onnx/onnx/commit/abb1702a)**: Add shape inference for Tile op (#2076) <Hariharan Seshadri>
- **[67638d9c](https://github.com/onnx/onnx/commit/67638d9c)**: [New Operator] Round (#2053) <Jeff Saremi>
- **[584e4477](https://github.com/onnx/onnx/commit/584e4477)**: Add dilations support in ConvTranspose shape inference and update docs (#2068) <daquexian>

Reviewed By: zrphercule

Differential Revision: D15762382

fbshipit-source-id: 590f25fb733e1565eb90fcdeb797b0ba34e2d2c3

jspark1105 committed:
StoreMatrixInMatrixMarketFormat can store both integer and float tensors (#21606)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21606

StoreMatrixInMatrixMarketFormat was able to dump quantized tensors only but sometimes we want to dump float tensors.

Reviewed By: csummersea

Differential Revision: D15741611

fbshipit-source-id: 95b03c2fdf1bd8407f7d925171d9dc9f25677464

mingzhe09088 committed:
Fix issue in backward path (#21663)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21663

as title

Reviewed By: hl475

Differential Revision: D15770793

fbshipit-source-id: b3d0dd030237c4d62bddc388984a273153fac4a6

Commit:
Emit Loop Condition as Separate Block (#21611)

Summary:
Emit loop condition as a separate block in loops, then inline them before conversion to SSA. This is needed for breaks & continues where we will inline the condition block after the continue pass and before the break pass.

I also considered emitting a prim::For and a prim::While, but i think it's easier to just have one pathway.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21611

Differential Revision: D15775820

Pulled By: eellison

fbshipit-source-id: de17c5e65f6e4a0256a660948b1eb630e41b04fb

Commit:
Override print when python is present (#21625)

Summary:
This makes it so we can see the output of prim::Print in environments like iPython notebooks which override sys.stdout
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21625

Differential Revision: D15756793

Pulled By: jamesr66a

fbshipit-source-id: 7d9a14b2e229ed358e784318e9d862677db2c461

jspark1105 committed:
correct measure quantization error when followed_by=Relu and dequantize_output=1 (#21664)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21664

As title

Reviewed By: csummersea

Differential Revision: D15770947

fbshipit-source-id: 57f5842e1a250300703b02134c314e4f06b767b8

ezyang committed:
Pin torchvision version. (#20811)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20811
ghimport-source-id: 52e043453272d8441a2c0efd7f005b71ded024d6

Differential Revision: D15779416

Pulled By: ezyang

fbshipit-source-id: 1b3c2d9aeab57e580038f0c2a8bfbfcae9d7b62a

mrshenli committed:
Delete DDP hooks in Reducer destructor (#21591)

Summary:
Closes https://github.com/pytorch/pytorch/issues/21344

DDP assigns the original module to the first module replica instead of creating a new one. Then, it creates a new Reducer to add post hooks to sync gradients. However, because every reconstructed DDP instance wraps the same original module, all their reducers will add hooks to the same set of variables. This PR deletes DDP hooks from variables when destructing Reducer, trying to make DDP failure recoverable.

pietern kuttas and I discussed the following solutions:

#### Solution 1

Keep `add_post_hook` API intact, and do a `dynamic_cast` in `del_post_hook` to check hook type. If the type matches Reducer's hook, delete it. As pietern mentioned, this will not work if we create multiple DDP instances from the same original model.

#### Solution 2

Use a counter to generate a unique key for every hook in `Function`, and keep them in a map. return the key to the caller of `add_post_hook`, and ask the caller to provide key if it needs to delete the hook.

Con: this would add extra overhead to `add_post_hook` and every `Function` object.

#### Solution 3 [Current implementation]

kuttas suggests that, instead of generating a unique key, directly using the address of the pointer would be better. In order to avoid messing up dereferencing, let `add_post_hook` to return a `uintptr_t`.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21591

Differential Revision: D15745706

Pulled By: mrshenli

fbshipit-source-id: e56d2d48de0c65f6667790ab16337eac7f7d8b76

Commit:
Back out "[pytorch][PR] [Re-landing] Fix caffe2 windows CI for new Windows AMI" (#21670)

Summary:
Original commit changeset: e65c1d6bfcc9
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21670

Differential Revision: D15776087

Pulled By: yf225

fbshipit-source-id: cbb55cbbcb133cae1aeb2fe75cc52e7350cc6c88

dskhudia committed:
Enable groupwise for qconv (#21592)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21592

We now support groupwise convolutions for qconv2d

Reviewed By: zafartahirov

Differential Revision: D15739239

fbshipit-source-id: 80b9b4fef5b9ee3d22ebecbaf205b970ab3d4250

daquexian committed:
Fix the shape of PReLU weight (#21330)

Summary:
Fix issue https://github.com/pytorch/pytorch/issues/21271
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21330

Reviewed By: zrphercule

Differential Revision: D15776459

Pulled By: houseroad

fbshipit-source-id: 4e0aef88e9c91c79faa3da6fa66f7466dee52018

chocjy committed:
Add div hashing

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21422

Reviewed By: xianjiec

Differential Revision: D15589181

fbshipit-source-id: f6ff0726164f88da45e4b090b4d5ad05305b3225

fehiepsi committed:
Add strong Wolfe line search for lbfgs (#8824)

Summary:
This pull request adds a line search for lbfgs. "strong Wolfe" is the default line search method in [minFunc](https://www.cs.ubc.ca/~schmidtm/Software/minFunc.html) and it is also recommended in the [Numerical Optimization](https://www.springer.com/gp/book/9780387303031) book.

The implementation is based on four sources:
+ https://www.cs.ubc.ca/~schmidtm/Software/minFunc.html
+ https://www.springer.com/gp/book/9780387303031 Algorithms 3.5, 3.6, formula 3.59
+ https://github.com/torch/optim/blob/master/lswolfe.lua
+ https://github.com/torch/optim/blob/master/polyinterp.lua

The 'lua' version is based on an old version of `minFunc`, which has been updated in 2012. I made a couple of small changes based on the updated version. Due to that, the test of comparing with `.lua` version is not consistent (that's is the reason I changed a learning rate in the test).
Pull Request resolved: https://github.com/pytorch/pytorch/pull/8824

Differential Revision: D15783067

Pulled By: vincentqb

fbshipit-source-id: 5316d9088233981120376d79c7869d5f97e51b69

mrshenli committed:
Fix ProcessGroupGloo allgather for tensors with shared storage (#21490)

Summary:
Fix https://github.com/pytorch/pytorch/issues/20421

`ProcessGroupGloo` only requires input/output tensors to be contiguous. Contiguous tensors might not start from the beginning of the underlying storage, e.g., `chunk(..., dim=0)[1]`. The current implementation passes `tensor.storage().data()` ptr to gloo buffer. This leads to wrong results if the tensor has a non-zero storage offset.

The proposed solution is to use `tensor.data_ptr()` instead. Let's see if this breaks any tests.

cc qijianan777
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21490

Differential Revision: D15768907

Pulled By: mrshenli

fbshipit-source-id: 9d7d1e9baf0461b31187c7d21a4a53b1fbb07397

Commit:
Add libtorch Linux CPU binary build to PR CI (#21671)

Summary:
Currently we don't have any Linux libtorch binary build in the PR CI, which led to nightly build failure such as https://circleci.com/gh/pytorch/pytorch/1939687. This PR adds Linux libtorch CPU binary build to prevent such breakage from happening in the future.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21671

Differential Revision: D15785003

Pulled By: yf225

fbshipit-source-id: d1f2e4235e48296ddecb3367f8e5a0df16f4ea49

Commit:
Updating submodules

Reviewed By: cdelahousse

fbshipit-source-id: 5cbf562652b9d7cf3877b5f819141f88c9b857d3

Commit:
nn.Transformer (#20170)

Summary:
Accidentally rebased the old PR and make it too messy. Find it here (https://github.com/pytorch/pytorch/pull/19274)

Create a PR for comments. The model is still WIP but I want to have some feedbacks before moving too far. The transformer model depends on several modules, like MultiheadAttention (landed).

Transformer is implemented based on the paper (https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf). Users have the flexibility to build a transformer with self-defined and/or built-in components (i.e encoder, decoder, encoder_layer, decoder_layer). Users could use Transformer class to build a standard transformer model and modify sub-layers as needed.

Add a few unit tests for the transformer module, as follow:
TestNN.test_Transformer_cell
TestNN.test_transformerencoderlayer
TestNN.test_transformerdecoderlayer
TestNN.test_transformer_args_check
TestScript.test_scriptmodule_transformer_cuda

There is another demonstration example for applying transformer module on the word language problem. https://github.com/pytorch/examples/pull/555
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20170

Differential Revision: D15417983

Pulled By: zhangguanheng66

fbshipit-source-id: 7ce771a7e27715acd9a23d60bf44917a90d1d572

zou3519 committed:
Add some more namedtensor builds to the CI (#21632)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21632
ghimport-source-id: 6a8da97ce153c6d279017af920edd0d20765c32c

Differential Revision: D15760331

Pulled By: zou3519

fbshipit-source-id: b2f4c65df5f6f9322d47da995c76851387e5df47

zou3519 committed:
Add some named tensor helper functions (#21636)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21636
ghimport-source-id: 5eff5744cd3c80f75bdb02576be1407a64e0434d

Differential Revision: D15780269

Pulled By: zou3519

fbshipit-source-id: 87ff40ffbe0ebd5fc4d105709c9f6f8dda5f9952

zou3519 committed:
Add virtual dtor to NamedTensorMetaInterface (#21633)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21633
ghimport-source-id: 6cdf0b1559e696a19e282ff6d5ba79c6b119e8c0

Differential Revision: D15760589

Pulled By: zou3519

fbshipit-source-id: 537882c05ab7b19889a31c648c5efeb1949831a8

Commit:
schema_matching.cpp: improve error messages.

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21141

Differential Revision: D15769066

Pulled By: ZolotukhinM

fbshipit-source-id: 5853e0360581c44e42b068add3bf2bc68e671b2b

syed-ahmed committed:
Refactor Random Number Generators in ATen (#21364)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21364
ghimport-source-id: ca7d37e10190ba46dc8512f437404ca9216d3369

Differential Revision: D15696497

Pulled By: ezyang

fbshipit-source-id: 2e713b8566ae915e175b5a79ac1dd9b86cc2a23d

smessmer committed:
Use c10::List (#21177)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21177

- Integrate c10::ListPtr into IValue and the c10 dispatcher.
- Streamline conversion to/from IValue. Before, we had IValue::to<> and kernel_functor.h had its own ivalue_to_arg_type and return_type_to_ivalue. They are now unified. Also, this means that nested types like Dicts of Lists of Optional of Dict of ... do work as expected now

Differential Revision: D15476433

fbshipit-source-id: bde9df80df20091aa8e6ae17ba7e90abd149b954

ZolotukhinM commented on torch/csrc/jit/passes/insert_guards.cpp:
```diff
@@ -33,14 +33,20 @@ struct GuardInserter {
+        } else {
+          // we didn't go down this path i.e
+          // no profiling information is available
+          n->output()->replaceAllUsesWith(n->input());
```
Where do we insert the guard in `else` condition?

Krovatkin replied:
we don't 😄  ... the if arm this `prim::profile` is in wasn't executed, so we don't have any profiling information to guard

ZolotukhinM commented on torch/csrc/jit/profiling_graph_executor_impl.cpp:
```diff
@@ -8,22 +11,34 @@ bool& getProfilingMode() {
+  return g;
+}
+
+ProfilingGraphExecutorImpl::ProfilingGraphExecutorImpl(const std::shared_ptr<Graph>& graph, bool optimize)
```
clang-format please :)

ZolotukhinM commented on torch/csrc/jit/profiling_graph_executor_impl.cpp:
```diff
@@ -8,22 +11,34 @@ bool& getProfilingMode() {
+
+  auto copy = pr_->graph()->copy();
+  InsertGuards(copy);
+  EliminateGuards(copy);
```
Nit: maybe it should be called `EliminateRedundantGuards` or something like that. Otherwise, it seems that these two lines should cancel each other:)

Krovatkin replied:
renamed it

ZolotukhinM reviewed (approved):
Generally looks good, but the code lacks comments IMHO. I pointed some of the places which were not obvious to me, but the list might not be exhaustive.

Apart from adding comments we might also do a better job at separating internal methods (i.e. marking them private) from public - even if the class/struct isn't supposed to be used anywhere else, such separation helps to understand the code.

Commit:
numerical integration "trapz" function. (#21610)

Summary:
This is intended to match [numpy.trapz](https://docs.scipy.org/doc/numpy/reference/generated/numpy.trapz.html): numerical integration based on the trapezoid rule.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21610

Differential Revision: D15747618

Pulled By: umanwizard

fbshipit-source-id: 8eadb2e75c9877b07592d875ca0b2cca6cb72297

Commit:
Support for type annotations instead of torch.jit.annotate() (#21390)

Summary:
This adds support for PEP 526 style annotations on assignments in place of
`torch.jit.annotate()`, so

```python
a = torch.jit.annotate(List[int], [])
```

turns into

```python
a : List[int] = []
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21390

Differential Revision: D15790937

Pulled By: driazati

fbshipit-source-id: 0cc204f7209a79839d330663cc6ba8320d3a4120

Commit:
Add `Final[T]` annotated members to `__constants__` (#21603)

Summary:
Class member annotations can be marked with `Final[T]` instead of adding them to `__constants__`. `Final` comes from the `typing_extensions` module (which will be used if it is present). If not, the polyfill from `_jit_internal` is exposed as `torch.jit.Final` for users that don't want to install `typing_extensions`.

This keeps around `__constants__` since a lot of code is still using it, but in documentation follow ups we should change the examples to all to use `Final`.

TODO: install typing_extensions on CI, move tests to a Python3 only file when #21489 lands
](https://our.intern.facebook.com/intern/diff/15746274/)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21603

Pulled By: driazati

Differential Revision: D15746274

fbshipit-source-id: d2c9b5643b4abba069b130c26fd42714c906ffac

Commit:
Add WeakIValue, use in tracer. (#21515)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21515
ghimport-source-id: 7898a68791db2b5050164ab01d6ca6991e05746d

Reviewed By: suo

Differential Revision: D15719981

Pulled By: zdevito

fbshipit-source-id: 42cf26cf6541bcdf95f1343da3b9228fe2c229da

zdevito committed:
make tests pass with enable_first_class_module() enabled. (#21565)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21565
ghimport-source-id: d1fe735fb7821eadc59116fb921d8fe39a49f818

Reviewed By: driazati

Differential Revision: D15729503

Pulled By: zdevito

fbshipit-source-id: fabb678f040d21fae7545e3b2be1d098e24c544e

zdevito committed:
Update quantization to work with first-class modules. (#21660)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21660
ghimport-source-id: f9a11b2748f49042ee636755358d79c547aa249e

Reviewed By: suo

Differential Revision: D15770237

Pulled By: zdevito

fbshipit-source-id: 41fa8577028eef247bc545635cd93192a0b19db4

Krovatkin committed:
Liveness for BailOut graphs

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21615

Differential Revision: D15793434

Pulled By: Krovatkin

fbshipit-source-id: d89f1bf61ea57a1e3b75f8e2b200c27beb8b46cf

Krovatkin committed:
Profiling Programs section in README.md

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21695

Differential Revision: D15795716

Pulled By: Krovatkin

fbshipit-source-id: e14a44210ea4312a247157a6681fce449e40f779

Krovatkin committed:
remove liveness tests

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21719

Differential Revision: D15797628

Pulled By: Krovatkin

fbshipit-source-id: 87742bdde0b05aff4341ababb1f55c51991768ec

kostmo committed:
Torch rename (#20774)

Summary:
This renames the CMake `caffe2` target to `torch`, as well as renaming `caffe2_gpu` to `torch_gpu` (and likewise for other gpu target variants).  Many intermediate variables that don't manifest as artifacts of the build remain for now with the "caffe2" name; a complete purge of `caffe2` from CMake variable names is beyond the scope of this PR.

The shell `libtorch` library that had been introduced as a stopgap in https://github.com/pytorch/pytorch/issues/17783 is again flattened in this PR.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20774

Differential Revision: D15769965

Pulled By: kostmo

fbshipit-source-id: b86e8c410099f90be0468e30176207d3ad40c821

Commit:
Revert D15769066: [pytorch][PR] schema_matching.cpp: improve error messages.

Differential Revision:
D15769066

Original commit changeset: 5853e0360581

fbshipit-source-id: ac6fa8429136abf4c7835919009f936eea11ea7b

XiaobingSuper committed:
Add mkldnn mul operator (#20575)

Summary:
### mkldnn backward ops list:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20567) Add aten mkldnn conv2d backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20570) Add aten mkldnn backward ops: relu, linear and reshape :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20571) Add aten mkldnn backward ops: max_pool2d, avg_pool2d and adaptive_avg_poo2d :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20572) Add aten mkldnn batchnorm backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20573) Add aten mkldnn zero_ operator:yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20575) Add mkldnn mul operator :yellow_heart:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20575

Differential Revision: D15799529

Pulled By: bddppq

fbshipit-source-id: 4887d8ef1a0e316ad9db199b657d9481fc13e486

vishwakftw committed:
Allow batch sizes > 65535 for inverse, solve, cholesky_solve and tria… (#21689)

Summary:
…ngular_solve

Changelog:
- Iterate over mini batches of 65535 matrices (maximum)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21689

Differential Revision: D15800254

Pulled By: soumith

fbshipit-source-id: c743ff13f1ba25d26874429d44e41a3c0ed21d6a

BowenBao committed:
Support tuples in ScriptModule inputs/outputs (#20784)

Summary:
- [x] Add tests after https://github.com/pytorch/pytorch/pull/20256 is merged

- Support exporting ScriptModule with inputs/outputs of arbitrarily constructed tuples.

- Moved the assigning of output shapes to after graph conversion to ONNX is completed. By then all tuples in the IR has already been lowered by the pass ```_jit_pass_lower_all_tuples```. If assigning output shapes is required to happen before that, we'll need to hand parse the tuple structures in the graph, and repeat the same logic in ```_jit_pass_lower_all_tuples```. Handling inputs is easier because all tuple information is encoded within the input tensor type.

- Swap the order of ```_jit_pass_lower_all_tuples``` and ```_jit_pass_erase_number_types```. Ops like ```prim::TupleIndex``` relies on index being a scalar. ```_jit_pass_erase_number_types``` will convert these kind of scalars to tensors.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20784

Reviewed By: zrphercule

Differential Revision: D15484171

Pulled By: houseroad

fbshipit-source-id: 4767a84038244c929f5662758047af6cb92228d3

JamesTrick committed:
Improving Categorical Distribution Docs' (#16291) (#21707)

Summary:
**Closes:** Confusing documentation with distributions.Categorical about logits https://github.com/pytorch/pytorch/issues/16291

**Solution**: Changes documentation on the Categorical distribution from `log probabilities` to `event log-odds`. This makes should reduce confusion as raised by this issue, and is consistent with other distributions such as `torch.Binomial`.

More than happy to make any other changes if they fit :).
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21707

Differential Revision: D15799181

Pulled By: soumith

fbshipit-source-id: f11acca7a5c130102a3ff6674640235ee5aa69bf

bddppq committed:
ROCm 2.5

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21724

Differential Revision: D15799149

Pulled By: bddppq

fbshipit-source-id: c72689e73470f2ca145556a2ac8cb34e36e341ef

xw285cornell committed:
Add use_rocm flag to detect AMD build in the runtime (#21718)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21718

adding a detection method on whether the package is built for AMD.

Reviewed By: bddppq

Differential Revision: D15795893

fbshipit-source-id: 91a21ee76b2273b1032507bdebe57e016717181d

skrah committed:
Port avg_pool2d() to ATen

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21635

Differential Revision: D15768487

Pulled By: ezyang

fbshipit-source-id: 85e1d883aded0f4d3ac5100719df335f5a337fc5

Commit:
Native TBB parallel backend (#20480)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20480
ghimport-source-id: c710f897c4c9b9616fc3dd76d80b4845aea43a1f

Differential Revision: D15333692

Pulled By: ilia-cher

fbshipit-source-id: 61e476dd5c737fe144e3aec000d8ebb11fbc0547

Commit:
Fix failing test for Final annotations (#21725)

Summary:
](https://our.intern.facebook.com/intern/diff/15800009/)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21725

Pulled By: driazati

Differential Revision: D15800009

fbshipit-source-id: 5409c213161e3f2031710933897b85872aad2a83

Commit:
at::launch benchmark (#21581)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21581
ghimport-source-id: 6a65d73694b17611d6ad45db0b39b86c318a68c7

Differential Revision: D15736495

Pulled By: ilia-cher

fbshipit-source-id: 6b9109ad3611ff3c8b1a37796e9149bef0c2ad36

Commit:
Fix first reshape in pixel_shuffle conversion (#21486)

Summary:
When converting pixel_shuffle to reshape + transpose + reshape, the first reshape should
be:
[N, C * r^2, H, W] => [N, C, r, r, H, W]
in order to match pytorch's implementation (see ATen PixelShuffle.cpp).

This previously wasn't caught by the test case, since it uses C = r = 4. Updated test case to
have C = 2, r = 4.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21486

Reviewed By: houseroad

Differential Revision: D15700945

Pulled By: houseroad

fbshipit-source-id: 47019691fdc20e152e867c7f6fd57da104a12948

nataliakliushkina committed:
Add event and event_counter columns to caffe2_usage_tracer table (#21739)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21739

Added event and event_counter columns for PyTorch/Caffe2 API usage metrics

Reviewed By: dzhulgakov

Differential Revision: D15119119

fbshipit-source-id: a71010bd659109a8e4f3a8bad84b22c1d15dc528

ssnl committed:
improve torch.load & torch.save doc formatting

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21747

Differential Revision: D15808189

Pulled By: ezyang

fbshipit-source-id: 5413eaaa901be098c6bad135f702ba103bc79d6c

Commit:
Skip failing test

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21751

Pulled By: driazati

Differential Revision: D15809091

fbshipit-source-id: 3cc96e632a7b89b4d86d68d2a76021d971447e12

Commit:
Add/edit docs for nn.transformer (#21746)

Summary:
Add docs for TransformerEncoder and TransformerDecoder, plus minor edits.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21746

Differential Revision: D15807498

Pulled By: zhangguanheng66

fbshipit-source-id: 388efb5821c4c3d25865cecea70902e9b2bf5d15

Commit:
add error message to missing function backend (#21742)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21742

Add error message to NotImplementedError so we know which function it is about.

Reviewed By: bddppq

Differential Revision: D15806379

fbshipit-source-id: 14eab9d03aa5b44ab95c5caeadc0e01d51f22188

izdeby committed:
Added dim check for index_copy_ (#21617)

Summary:
Fixing reported [bug](https://github.com/pytorch/pytorch/issues/20322)
The issue was related to not checking the dimensions of source vs destination tensors.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21617

Differential Revision: D15749963

Pulled By: izdeby

fbshipit-source-id: acff114c729fd9c0a9a51325e0ebd8b42e1f2fc1

Krovatkin commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -879,7 +949,28 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+                af.functions[inst.X]->get_executor().getPlanFor(stack).code;
+            size_t num_inputs = code.num_inputs();
+            size_t base_pointer = frames.back().base_pointer;
+            size_t inputs_start = stack.size() - num_inputs;
```
fixed

BowenBao committed:
Bug fix: ONNX export full operator (#21669)

Summary:
Fix an obvious bug.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21669

Reviewed By: zrphercule

Differential Revision: D15806614

Pulled By: houseroad

fbshipit-source-id: d0f6e934252e0057f3dbcc7f160236ee6f4497ac

Krovatkin commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -587,6 +620,36 @@ struct CodeImpl {
+  void emitBailOut(Node* node) {
+    emitLoadInputs(node->inputs().slice(0, 1));
+    insertInstruction(GUARD, type_table_.size());
+    type_table_.emplace_back(node->outputs().at(0)->type());
```
switched to `output()` which asserts this exact invariant

Krovatkin commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -688,6 +754,7 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   struct Frame {
     std::shared_ptr<CodeImpl> function;
     size_t pc;
+    size_t base_pointer;
```
added a comment

Krovatkin commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -384,6 +387,11 @@ struct WithCurrentNode {
   Node* old_value_;
 };
 
+struct BailoutBlock {
```
added a comment

Krovatkin commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -587,6 +620,36 @@ struct CodeImpl {
+    insertInstruction(JF, 0 /* to be patched */);
+    size_t jf_index = instructions_.size() - 1;
+    emitLoadInputs(node->inputs().slice(1));
+    insertInstruction(TAIL_CALL, function_table_.size());
```
I can only speculate here. I guess @zdevito might think that `function_table_.size() - 1` is less tidy?

ZolotukhinM replied:
It's minor, so I don't mind either way.

ezyang committed:
Document that no_grad is thread local. (#21755)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21755
ghimport-source-id: dfb53759024d9ba9d104fdb2a8151ab996e55234

Differential Revision: D15811172

Pulled By: ezyang

fbshipit-source-id: c8c7c1c15277d8fe8cc513e20af449257d7ff15c

Commit:
Use expect to fill in pytorchbot token (#20459)

Summary:
In this PR, we use `expect` to fill in the token for pytorchbot when doing `git push`, so that we don't need to save the token in the git remote URL.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20459

Differential Revision: D15811676

Pulled By: yf225

fbshipit-source-id: cd3b780da05d202305f76878e55c3435590f15a8

Commit:
fix multihead attention for half (#21658)

Summary:
Currently multihead attention for half type is broken
```
  File "/home/ngimel/pytorch/torch/nn/functional.py", line 3279, in multi_head_attention_forward
    attn_output = torch.bmm(attn_output_weights, v)
RuntimeError: Expected object of scalar type Float but got scalar type Half for argument https://github.com/pytorch/pytorch/issues/2 'mat2'
```
because softmax converts half inputs into fp32 inputs. This is unnecessary - all the computations in softmax will be done in fp32 anyway, and the results need to be converted into fp16 for the subsequent batch matrix multiply, so nothing is gained by writing them out in fp32. This PR gets rid of type casting in softmax, so that half works.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21658

Differential Revision: D15807487

Pulled By: zhangguanheng66

fbshipit-source-id: 4709ec71a36383d0d35a8f01021e12e22b94992d

chosungmann committed:
Fix spelling errors (#21665)

Summary:
alloctor -> allocator
excutable -> executable
excution -> execution
foward -> forward
initiaize -> initialize
paralell -> parallel
preprocesor -> preprocessor
tranpose -> transpose
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21665

Differential Revision: D15806155

Pulled By: soumith

fbshipit-source-id: d92b21ec8650a2b32f05faf9af0b7d2b073e992c

colesbury committed:
Tune elementwise ops for ROCm (#21754)

Summary:
```
The stride calculation using OffsetCalculator performs poorly with
MAX_DIMS=25. This reduces MAX_DIMS (after coalescing) to 16 on ROCm.
I think it's unlikely that anyone will exceed this limit. If they do,
we can add additional specializations for ROCm with more dimensions.
```

I'm not sure about the underlying cause. With MAX_DIM=25, the add kernel's params
is ~648 bytes vs. ~424 bytes with MAX_DIM=16. The kernel instruction footprint is
bigger too, but most of these instructions are never executed and most kernel parameters
are never loaded because the typical dimensionality is much smaller.

Mini benchmark here:
https://gist.github.com/colesbury/1e917ae6a0ca9d24712121b92fed4c8f

(broadcasting operations are much faster)

cc iotamudelta
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21754

Reviewed By: bddppq

Differential Revision: D15811906

Pulled By: colesbury

fbshipit-source-id: 063f92c083d26e2ef2edc98df7ff0400f9432b9d

Commit:
schema_matching.cpp: improve error messages.

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/21141

Differential Revision: D15808354

Pulled By: ZolotukhinM

fbshipit-source-id: 16d938fd5acafb445a0c433cabc9a55cab563165

Commit:
Add batch id to tracer event (#21446)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21446

this is used for easier tracing of iter id when looking at trace diagram

Reviewed By: ilia-cher

Differential Revision: D15628950

fbshipit-source-id: ee75b3bdb14a36abc18c7bddc49d8ec9789b724d

Commit:
better example for local weights (#21685)

Summary:
fixes https://github.com/pytorch/hub/issues/29
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21685

Differential Revision: D15817774

Pulled By: ailzhang

fbshipit-source-id: d2f615e5d431186d45a21d8300fb9ba3c37b246c

ezyang committed:
Make it possible to trigger all tests by pushing to ci-all/ branch (#21750)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21750
ghimport-source-id: 4792aa5ccab7e4b54c21f23d0b78802f85bbeb8d

Differential Revision: D15819367

Pulled By: ezyang

fbshipit-source-id: db91ee727c66469ac78e59b3662f29db53a916bc

zou3519 committed:
Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten (#21647)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21647
ghimport-source-id: 1db4ec31f047f7854a39c28e2b38918dc6b44f42

Differential Revision: D15804425

Pulled By: zou3519

fbshipit-source-id: 575cc3de09287efe75e7052df129626748208d0d

anshuljain1 committed:
Handling numel() == 0 in convTranspose (#21652)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21652

Diff fixes issue of empty ROIs for convTranspose

Issue StackTrace: P65374505

Reviewed By: jerryzh168

Differential Revision: D15766739

fbshipit-source-id: 39cf8feca66b6aae22ff4ec5c1b6a4e3f20f378d

Commit:
backend.py: _getattr__ must raise AttributeError (#21763)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21763

Custom __getattr__ functions can only raise AttributeError. This code throwed NotImplementedError which caused upstream troubles when hasattr() was called.

Differential Revision: D15815176

fbshipit-source-id: 0982e2382de4578d3fc05c5d2a63f624d6b4765e

Commit:
Query caffe2 operator stats for detailed execution info (#20924)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20924

I found a python3 bug for deserializing caffe2 code. The exception thrown is Unicode related error instead of just decode error, and we need to catch that as well

Reviewed By: ipiszy

Differential Revision: D15293221

fbshipit-source-id: 29820800d1b4cbe5bf3f5a189fe2023e655d0508

shagunsodhani committed:
Fix the issue of sizes vs size for tensor creation ops (#21686)

Summary:
Related to [pytorch#20921](https://github.com/pytorch/pytorch/issues/20921)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21686

Differential Revision: D15816109

Pulled By: gchanan

fbshipit-source-id: 4428b8e77b6c8b297ddb77e58fc1cb916c9cc46e

definitelynotmcarilli committed:
Fix with emit_nvtx, also allow shape information to appear in nvtx ranges. (#21691)

Summary:
This PR is intended as a fix for https://github.com/pytorch/pytorch/issues/21644.

It allows the `with emit_nvtx` context manager to take an additional `record_shapes` argument. `record_shapes` is False by default, but if True, the nvtx ranges generated for each autograd op will append additional information about the sizes of Tensors received by that op.

The format of shape information is equivalent to what the CPU-side profiler spits out.  For example,
```
M = torch.randn(2, 3)
mat1 = torch.randn(2, 3)
mat2 = torch.randn(3, 3)

with torch.cuda.profiler.profile():
    with torch.autograd.profiler.emit_nvtx(record_shapes=True):
        torch.addmm(M, mat1, mat2)
```
produces the following nvtx range label for addmm:
![Screenshot from 2019-06-12 10-48-01](https://user-images.githubusercontent.com/7799218/59374008-b7d13100-8cff-11e9-9245-58410073d965.png)
(cf the "Input Shapes" shown in https://github.com/pytorch/pytorch/commit/864cfbc2162a874fd67b50414205483cb9ac6b5d#diff-115b6d48fa8c0ff33fa94b8fce8877b6)

I also took the opportunity to do some minor docstring cleanup.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21691

Differential Revision: D15816226

Pulled By: gchanan

fbshipit-source-id: b2b01ea10fea61a6409a32b41e85b6c8b4851bed

xuhdev committed:
When converting to NumPy, throw TypeError when type is not supported (#21608)

Summary:
This makes the error thrown in aten_to_numpy_dtype consistent with that in numpy_dtype_to_aten.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21608

Differential Revision: D15816035

Pulled By: gchanan

fbshipit-source-id: 392e8b9ea37003a859e7ed459911a1700fcbd695

xuhdev committed:
When building extensions, honor options set in CMake. (#21653)

Summary:
Currently when building extensions, variables such as USE_CUDA, USE_CUDNN are used to determine what libraries should be linked. But we should use what CMake has detected, because:

1. If CMake found them unavailable but the variables say some libraries should be linked, the build would fail.
2. If the first build is made using a set of non-default build options, rebuild must have these option passed to setup.py again, otherwise the extension build process is inconsistent with CMake. For example,

```bash
# First build
USE_CUDA=0 python setup.py install
# Subsequent builds like this would fail, unless "build/" is deleted
python setup.py install
```

This commit addresses the above issues by using variables from CMakeCache.txt when building the extensions.

 ---

The changes in `setup.py` may look lengthy, but the biggest changed block is mostly moving them into a function `configure_extension_build` (along with some variable names changed to `cmake_cache_vars['variable name']` and other minor changes), because it must be called after CMake has been called (and thus the options used and system environment detected by CMake become available).
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21653

Differential Revision: D15824506

Pulled By: ezyang

fbshipit-source-id: 1e1eb7eec7debba30738f65472ccad966ee74028

LeviViana committed:
Throwing errors for min and max reductions in empty CUDA tensors (#19612)

Summary:
Related to https://github.com/pytorch/pytorch/issues/17750.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19612

Differential Revision: D15813649

Pulled By: gchanan

fbshipit-source-id: aa3dc34dd1e6d8bb24fa4c18891204108759bb35

ssnl committed:
rebuild_storage_fd retry on EINTR (#21723)

Summary:
Some data loader tests are flaky on py 2 with the following error
```
Jun 12 22:17:31 Traceback (most recent call last):
Jun 12 22:17:31   File "test_dataloader.py", line 798, in test_iterable_dataset
Jun 12 22:17:31     fetched = sorted([d.item() for d in dataloader_iter])
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/utils/data/dataloader.py", line 697, in __next__
Jun 12 22:17:31     idx, data = self._get_data()
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/utils/data/dataloader.py", line 664, in _get_data
Jun 12 22:17:31     success, data = self._try_get_data()
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/utils/data/dataloader.py", line 617, in _try_get_data
Jun 12 22:17:31     data = self.data_queue.get(timeout=timeout)
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/multiprocessing/queues.py", line 135, in get
Jun 12 22:17:31     res = self._recv()
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/multiprocessing/queue.py", line 22, in recv
Jun 12 22:17:31     return pickle.loads(buf)
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/pickle.py", line 1382, in loads
Jun 12 22:17:31     return Unpickler(file).load()
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/pickle.py", line 858, in load
Jun 12 22:17:31     dispatch[key](self)
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/pickle.py", line 1133, in load_reduce
Jun 12 22:17:31     value = func(*args)
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/multiprocessing/reductions.py", line 274, in rebuild_storage_fd
Jun 12 22:17:31     fd = multiprocessing.reduction.rebuild_handle(df)
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/multiprocessing/reduction.py", line 157, in rebuild_handle
Jun 12 22:17:31     new_handle = recv_handle(conn)
Jun 12 22:17:31   File "/opt/python/2.7.9/lib/python2.7/multiprocessing/reduction.py", line 83, in recv_handle
Jun 12 22:17:31     return _multiprocessing.recvfd(conn.fileno())
Jun 12 22:17:31 OSError: [Errno 4] Interrupted system call
```

Apparently, Python 2.7's `recvfd` calls `recvmsg` without EINTR retry: https://github.com/python/cpython/blob/2.7/Modules/_multiprocessing/multiprocessing.c#L174
So we should call it with an outer try-catch loop.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21723

Differential Revision: D15806247

Pulled By: ezyang

fbshipit-source-id: 16cb661cc0fb418fd37353a1fef7ceeb634f02b7

XiaobingSuper committed:
Add aten mkldnn zero_ operator (#20573)

Summary:
### mkldnn backward ops list:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20567) Add aten mkldnn conv2d backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20570) Add aten mkldnn backward ops: relu, linear and reshape :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20571) Add aten mkldnn backward ops: max_pool2d, avg_pool2d and adaptive_avg_poo2d :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20572) Add aten mkldnn batchnorm backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20573) Add aten mkldnn zero_ operator:yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20575) Add mkldnn mul operator 💚
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20573

Differential Revision: D15820477

Pulled By: bddppq

fbshipit-source-id: 35d95f5b4e013c8db1911f52148550a2e40a2e68

houseroad committed:
Add ONNX py3 CI cases (#21715)

Summary:
So far, we only have py2 ci for onnx. I think py3 support is important. And we have the plan to add onnxruntime backend tests, which only supports py3.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21715

Reviewed By: bddppq

Differential Revision: D15796885

Pulled By: houseroad

fbshipit-source-id: 8554dbb75d13c57b67ca054446a13a016983326c

nairbv committed:
Fix handling of kwargs from common method invocations (#21499)

Summary:
When kwargs are specified in a test defined via common_method_invocations, it doesn't work if there isn't also a positional argument (`{'foo':'foo'}` without a positional arg generates a python call like: `self.method(, foo=foo)`, erroring on the `,`). I wanted to test something in a different PR and noticed I couldn't.

Also fixed some flake8 warnings I was seeing locally.

I replaced `lambda x: x` with `ident` since it seems a bit cleaner to me, but happy to revert that if others don't agree?
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21499

Differential Revision: D15826974

Pulled By: nairbv

fbshipit-source-id: a3f37c80ba2303c7d9ae06241df06c7475b64e36

ezyang committed:
Define __setstate__ on _ConvNd to handle pre-padding_mode pickles. (#21687)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21687
ghimport-source-id: df49530d25239ac4d62eae83c5d7b0d8f00f836a

Differential Revision: D15807402

Pulled By: ezyang

fbshipit-source-id: f51b221444afc4e017db7544642a9c0a7d2a3efb

Krovatkin commented:
rebased to target master in PR #21799
